### PR TITLE
base-files: /etc/profile service improvement

### PR DESCRIPTION
**Pull request by stangri**
This patch introduces: 
1. ```service enabled``` to list the status (enabled/disabled) of services in ```/etc/init.d```
2. ```service <service-name-here> ubus``` to display the ```ubus call service list``` output for specified service

Signed-off-by: Stan Grishin <stangri@melmac.net>

### CODE REVIEW

**Comment by adrianschmutzler** [comment]
Please add the description to the commit message.

**Comment by stangri** [comment]
> Please add the description to the commit message.

done.

**Comment by adrianschmutzler** [comment]
Description has to be _before_ signed-off, so signed-off is the last line.

**Comment by ysc3839** [comment]
Why not implement `ubus` command in `/etc/rc.common`?

**Comment by stangri** [comment]
> Description has to be _before_ signed-off, so signed-off is the last line.

Done.
PS. If it's that important, I think it should be added to the submitting patches text, rather than hotlinking that instruction.

**Comment by stangri** [comment]
> Why not implement `ubus` command in `/etc/rc.common`?

I'm not sure if that's the best approach.

File: package/base-files/files/etc/profile [hunk 1]
```diff
@@ -49,11 +49,18 @@ in order to prevent unauthorized SSH logins.
 EOF
 fi
 
-service() {
-	[ -f "/etc/init.d/$1" ] || {
+service() { 
+	if [ "$1" = "enabled" ] && [ -z "$2" ]; then
```
Comment (ynezz) [inline package/base-files/files/etc/profile]: Seems confusing to me, this is going to be called as `service enabled` and list enabled AND disabled services. 

Perhaps something like `service show-enabled` (or `list-enabled`) and `service show-disabled` (or `list-disabled`) would make more sense?

Comment (stangri) [reply package/base-files/files/etc/profile]: Oh, great idea, perhaps list-enabled to be consistent with opkg's list-installed?

Unless there are other comments, I'll implement the fix and push it later this week.

Comment (stangri) [reply package/base-files/files/etc/profile]: Implemented list-enabled and list-disabled as per our conversation.

**Comment by stangri** [comment]
Implemented the list-enabled and list-disabled as per our earlier conversation. Should I make a separate commit for 19.07 an 18.06?

**Comment by adrianschmutzler** [comment]
@stangri Let's focus on master for know. To me this looks like a backport would be trivial, so we wouldn't need a commit for it.

The question to me is whether this should be backported at all, as it looks like a feature to me (rather than a fix).

**Comment by ynezz** [comment]
> whether this should be backported

Nope.

File: package/base-files/files/etc/profile [hunk 2]
```diff
@@ -49,11 +49,22 @@ in order to prevent unauthorized SSH logins.
...
+	elif [ "$1" = "list-disabled" ] && [ -z "$2" ]; then 
+		for F in /etc/init.d/* ; do 
+			$F enabled || echo "$F disabled"
+		done; 
+	elif [ -f "/etc/init.d/$1" ] && [ "$2" = "ubus" ]; then
+		ubus call service list "{'name': '$1'}"
+	elif [ -f "/etc/init.d/$1" ]; then 
+		/etc/init.d/$@
+	else 
 		echo "service "'"'"$1"'"'" not found, the following services are available:"
```
Comment (wulfy23) [inline package/base-files/files/etc/profile]: echo "service list-enabled|list-disabled|<name start|stop|restart|reload|enable|disable|ubus>" ... etc ...

Comment (stangri) [reply package/base-files/files/etc/profile]: There could be more commands beyond listed, so my fix is slightly different from suggested.

**Comment by stangri** [comment]
> The question to me is whether this should be backported at all, as it looks like a feature to me (rather than a fix).

19.07 hasn't been released yet, I would strongly suggest it's backported to 19.07.

**Comment by stangri** [comment]
@adrianschmutzler @ynezz -- Should the "needs changes" tag now be removed? What's preventing this from being merged?

**Comment by stangri** [comment]
@wulfy23 @ysc3839 any more feedback?

**Comment by oldium** [comment]
Regarding ubus argument - from the architecture point of view simulating/stealing ubus argument in context, which is usually just forwarded to init script, is not good. For me logical would be to be able to call /etc/init.d/someservice ubus, so I would implement it in rc.common too.

**Comment by stangri** [comment]
> Regarding ubus argument - from the architecture point of view simulating/stealing ubus argument in context, which is usually just forwarded to init script, is not good. For me logical would be to be able to call /etc/init.d/someservice ubus, so I would implement it in rc.common too.

Moving that ubus call to /etc/rc.common leads to an indefinite loop if I do this:
```sh
ubus() {
	name="$(basename "${initscript}")"
#	echo "ubus - $name"
	ubus call service list "{'name': '$name'}"
	return 0
}

ALL_COMMANDS="start stop reload restart boot shutdown enable disable enabled depends ubus ${EXTRA_COMMANDS}"
list_contains ALL_COMMANDS "$action" || action=help
$action "$@"
```

If you have suggestions on how to overcome this, please let me know.

**Comment by oldium** [comment]
`ubus` is not very intuitive anyway, so I would rename `ubus` to something like `status-ubus` or enhance the current `status` by optional parameter `--ubus`.

**Comment by stangri** [comment]
> `ubus` is not very intuitive anyway, so I would rename `ubus` command to something like `status-ubus` or enhance the current `status` by optional parameter `--ubus`.

Sure, I can rename it to ubusinfo or whatever. The issue is that calling ubus from rc.common causes the loop. You know how to overcome this?

**Comment by oldium** [comment]
Sorry, just updated my latest comment - please see update.

**Comment by oldium** [comment]
@stangri Issue with loop is caused by the fact that you have function `ubus()`, which is called in the line `ubus call ...`. When you rename `ubus()` to something else, the loop disappears.

**Comment by stangri** [comment]
> Sorry, just updated my latest comment - please see update.

We're on the same page with the name and I can address it later. 

> @stangri Issue with loop is caused by the fact that you have function `ubus()`, which is called in the line `ubus call ...`. When you rename `ubus()` to something else, the loop disappears.

OMG, my only excuse that it's 4:30am on the NY Day! Just replaced the call with `/bin/ubus` and the problem went away.

**Comment by stangri** [comment]
Renamed `ubus` to `status-ubus` and pushed fixes to `/etc/profile` and `/etc/rc.common`. I still think `ubus` is a better name as it would have allowed to keep the same tabbing in the `help()` function.

**Comment by stangri** [comment]
Any feedback?

**Comment by ynezz** [comment]
Please rebase properly and `git push -f` in order to get rid of that additional merge commit.

**Comment by ynezz** [comment]
While at it line wrap your commit description to < 72 chars.

**Comment by stangri** [comment]
Thanks, I'll submit thru mail list as per your instructions.

**Pull request closed without merging**